Random Forest - Base

simarjeet30 opened the pull request.

ashwani-rathee commented:
@simarjeet30 The algorithm needs to be written from scratch with numpy only,not using sklearn :)
Pls update the PR accordingly

ashwani-rathee commented on Random Forest/random_forest/random_forest.py:
```diff
@@ -0,0 +1,24 @@
+sc = StandardScaler()
+X_train = sc.fit_transform(X_train)
+X_test = sc.transform(X_test)
+from sklearn.ensemble import RandomForestRegressor
```
You are not supposed to import Random Forest but implement it :)

simarjeet30 commented:
Ok I have tried again with other data ,i want to show U first before
updating the PR. please take a look at it .


class RandomForest():
def __init__(self, x, y, n_trees, n_features, sample_sz, depth=10, min_leaf=
5):
np.random.seed(12)
if n_features == 'sqrt':
self.n_features = int(np.sqrt(x.shape[1]))
elif n_features == 'log2':
self.n_features = int(np.log2(x.shape[1]))
else:
self.n_features = n_features
print(self.n_features, "sha: ",x.shape[1])
self.x, self.y, self.sample_sz, self.depth, self.min_leaf = x, y, sample_sz,
depth, min_leaf
self.trees = [self.create_tree() for i in range(n_trees)]
def create_tree(self):
idxs = np.random.permutation(len(self.y))[:self.sample_sz]
f_idxs = np.random.permutation(self.x.shape[1])[:self.n_features]
return DecisionTree(self.x.iloc[idxs], self.y[idxs], self.n_features, f_idxs
,
idxs=np.array(range(self.sample_sz)),depth = self.depth, min_leaf=self.
min_leaf)
def predict(self, x):
return np.mean([t.predict(x) for t in self.trees], axis=0)
def std_agg(cnt, s1, s2): return math.sqrt((s2/cnt) - (s1/cnt)**2)
class DecisionTree():
def __init__(self, x, y, n_features, f_idxs,idxs,depth=10, min_leaf=5):
self.x, self.y, self.idxs, self.min_leaf, self.f_idxs = x, y, idxs, min_leaf,
f_idxs
self.depth = depth
print(f_idxs)
# print(self.depth)
self.n_features = n_features
self.n, self.c = len(idxs), x.shape[1]
self.val = np.mean(y[idxs])
self.score = float('inf')
self.find_varsplit()
def find_varsplit(self):
for i in self.f_idxs: self.find_better_split(i)
if self.is_leaf: return
x = self.split_col
lhs = np.nonzero(x<=self.split)[0]
rhs = np.nonzero(x>self.split)[0]
lf_idxs = np.random.permutation(self.x.shape[1])[:self.n_features]
rf_idxs = np.random.permutation(self.x.shape[1])[:self.n_features]
self.lhs = DecisionTree(self.x, self.y, self.n_features, lf_idxs, self.idxs[
lhs], depth=self.depth-1, min_leaf=self.min_leaf)
self.rhs = DecisionTree(self.x, self.y, self.n_features, rf_idxs, self.idxs[
rhs], depth=self.depth-1, min_leaf=self.min_leaf)
def find_better_split(self, var_idx):
x, y = self.x.values[self.idxs,var_idx], self.y[self.idxs]
sort_idx = np.argsort(x)
sort_y,sort_x = y[sort_idx], x[sort_idx]
rhs_cnt,rhs_sum,rhs_sum2 = self.n, sort_y.sum(), (sort_y**2).sum()
lhs_cnt,lhs_sum,lhs_sum2 = 0,0.,0.
for i in range(0,self.n-self.min_leaf-1):
xi,yi = sort_x[i],sort_y[i]
lhs_cnt += 1; rhs_cnt -= 1
lhs_sum += yi; rhs_sum -= yi
lhs_sum2 += yi**2; rhs_sum2 -= yi**2
if i<self.min_leaf or xi==sort_x[i+1]:
continue
lhs_std = std_agg(lhs_cnt, lhs_sum, lhs_sum2)
rhs_std = std_agg(rhs_cnt, rhs_sum, rhs_sum2)
curr_score = lhs_std*lhs_cnt + rhs_std*rhs_cnt
if curr_score<self.score:
self.var_idx,self.score,self.split = var_idx,curr_score,xi
@property
def split_name(self): return self.x.columns[self.var_idx]
@property
def split_col(self): return self.x.values[self.idxs,self.var_idx]
@property
def is_leaf(self): return self.score == float('inf') or self.depth <= 0
def predict(self, x):
return np.array([self.predict_row(xi) for xi in x])
def predict_row(self, xi):
if self.is_leaf: return self.val
t = self.lhs if xi[self.var_idx]<=self.split else self.rhs
return t.predict_row(xi)


On Sun, Mar 14, 2021 at 1:23 PM Ashwani Rathee ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In Random Forest/random_forest/random_forest.py
> <https://github.com/Algo-Phantoms/Algo-ScriptML/pull/76#discussion_r593858859>
> :
>
> > @@ -0,0 +1,24 @@
> +import pandas as pd
> +import numpy as np
> +dataset = pd.read_csv("D:/Datasets/bill_authentication.csv")
> +dataset.head()
> +X = dataset.iloc[:, 0:4].values
> +y = dataset.iloc[:, 4].values
> +from sklearn.model_selection import train_test_split
> +
> +X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
> +from sklearn.preprocessing import StandardScaler
> +
> +sc = StandardScaler()
> +X_train = sc.fit_transform(X_train)
> +X_test = sc.transform(X_test)
> +from sklearn.ensemble import RandomForestRegressor
>
> You are not supposed to import Random Forest but implement it :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Algo-Phantoms/Algo-ScriptML/pull/76#pullrequestreview-611668530>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASZVY2B7E4INXBN4FPIPEIDTDRTPXANCNFSM4ZEVFA3Q>
> .
>

ashwani-rathee commented:
@simarjeet30 This looks like what I'm looking for but pls send it in PR form with proper indentation,it's easier to review that..

simarjeet30 commented:
Ok thanks for reviewing ,i will do that just now.

On Mon 15 Mar, 2021, 8:37 AM Ashwani Rathee, ***@***.***>
wrote:

> @simarjeet30 <https://github.com/simarjeet30> This looks like what I'm
> looking for but pls send it in PR form with proper indentation,it's easier
> to review that..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Algo-Phantoms/Algo-ScriptML/pull/76#issuecomment-799058342>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASZVY2CDUOTOGEV62ULSYGTTDV2W5ANCNFSM4ZEVFA3Q>
> .
>

simarjeet30 commented:
I have made changes to the same pull request .

On Mon, Mar 15, 2021 at 8:40 AM simarjeet bhatia ***@***.***>
wrote:

> Ok thanks for reviewing ,i will do that just now.
>
> On Mon 15 Mar, 2021, 8:37 AM Ashwani Rathee, ***@***.***>
> wrote:
>
>> @simarjeet30 <https://github.com/simarjeet30> This looks like what I'm
>> looking for but pls send it in PR form with proper indentation,it's easier
>> to review that..
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Algo-Phantoms/Algo-ScriptML/pull/76#issuecomment-799058342>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ASZVY2CDUOTOGEV62ULSYGTTDV2W5ANCNFSM4ZEVFA3Q>
>> .
>>
>

ashwani-rathee commented on Random Forest/random_forest/random_forest.py:
```diff
@@ -1,24 +1,85 @@
-import pandas as pd
```
You use lot of numpy functions in the random forest class but you deleted out numpy import.Please import numpy properly..

ashwani-rathee replied:
#64 #11 #44 are good examples of proper implementation by your fellow participants

ashwani-rathee commented:
@simarjeet30 Hey,I tried to run this but I am unable to.I understand that as first time contributor,you wish to get this merged fast but it has to run and give me some results which is not yet able,can you document it better?...so that I know how to run this properly later,as I tried it with bunch of parameters today with those mentioned in code but it still doesn't run
Several Good examples that will help you understand random forest better:
- https://machinelearningmastery.com/implement-random-forest-scratch-python/
- https://towardsdatascience.com/random-forests-and-decision-trees-from-scratch-in-python-3e4fa5ae4249
- https://towardsdatascience.com/building-a-random-forest-classifier-c73a4cae6781
- https://www.analyticsvidhya.com/blog/2018/12/building-a-random-forest-from-scratch-understanding-real-world-data-products-ml-for-programmers-part-3/

ashwani-rathee commented:
@simarjeet30 This stands invalid since it's plagiarized completely from https://towardsdatascience.com/random-forests-and-decision-trees-from-scratch-in-python-3e4fa5ae4249 .@geekquad

ashwani-rathee commented:
@simarjeet30 Plagiarized
